coordinator/src/orderbook/mod.rs

```diff
@@ -30,120 +41,317 @@ pub mod websocket;
 #[cfg(test)]
 mod tests;
 
-pub struct Orderbook {
+struct Orderbook {
     pool: Pool<ConnectionManager<PgConnection>>,
-    shorts: Shorts,
-    longs: Longs,
+    shorts: OrderbookSide,
+    longs: OrderbookSide,
+    // TODO(holzeis): Split up order matching fee rate into taker and maker fees.
+    order_matching_fee_rate: Decimal,
 }
 
-pub struct Shorts(Vec<Order>);
+struct OrderbookSide {
+    orders: Vec<Order>,
+}
 
-impl Shorts {
-    pub fn add(&mut self, order: Order) {
-        self.0.push(order);
+impl OrderbookSide {
+    pub fn new() -> Self {
+        Self { orders: vec![] }
     }
-    pub fn remove(&mut self, order_id: Uuid) {
-        self.0.retain(|o| o.id != order_id)
+    /// adds the given order to the orderbook.
+    pub fn add_order(&mut self, order: Order) {
+        self.orders.push(order);
+        self.sort(order.direction);
     }
-    pub fn sort(&mut self) {
-        self.0.sort_by(|a, b| {
-            if a.price.cmp(&b.price) == Ordering::Equal {
-                return a.timestamp.cmp(&b.timestamp);
-            }
-            // Descending order.
-            b.price.cmp(&a.price)
-        })
+    /// updates an order in the orderbook. Ignores the update if order does not exist.
+    pub fn update_order(&mut self, updated_order: Order) {
+        if let Some(order) = self.orders.iter_mut().find(|o| o.id == updated_order.id) {
+            *order = updated_order;
+        }
+        self.sort(updated_order.direction);
     }
-}
+    /// removes the order by the given id from the orderbook.
+    pub fn remove_order(&mut self, order_id: Uuid) {
+        self.orders.retain(|o| o.id != order_id);
+    }
+    /// removes the first order from the sorted list representing the best order in the orderbook.
+    pub fn take_order(&mut self) -> Option<Order> {
+        match self.orders.is_empty() {
+            true => None,
+            false => Some(self.orders.remove(0)),
+        }
+    }
+    /// matches orders from the orderbook for the given quantity.
+    pub fn match_order(&mut self, quantity: Decimal) -> Vec<Order> {
+        let mut matched_orders = vec![];
+
+        let mut quantity = quantity;
+        while let Some(mut order) = self.take_order() {
```

Comment:
🤔 I'm not a big fan of mutating the `OrderbookSide` before you know if you are gonna match the orders or not. This leads to having to roll back your work afterwards.

I think we should instead iterate through the top of the orderbook side and only take the orders after we have confirmed that we should.

Reply:
I've added a `commit_matched_orders` function, that only removes (takes) the orders from the orderbook if the quantity is enough.